File: ATG/app/src/main/java/com/kuhakupixel/atg/backend/ACEBaseClient.java
package com.kuhakupixel.atg.backend;

import android.util.Log;

import java.util.List;

public abstract class ACEBaseClient {

    public static class InvalidCommandException extends RuntimeException {
        public InvalidCommandException(String msg) {
            super(msg);
        }
    }

    public ACEBaseClient() {
    }

    public static void AssertValidCommand(List<String> out) throws InvalidCommandException {
        if (out.size() > 0) {
            if (out.get(0).equals("INVALID_COMMAND"))
                throw new InvalidCommandException("Invalid Command: " + String.join("\n", out));
        }
    }

    /**
     * specify how to send the command and get the output
     * */
    public abstract List<String> SendCommand(String[] requestCmd);

    public List<String> RequestAsList(String[] requestCmd) throws InvalidCommandException {

        Log.i("ATG", String.format("Command to Engine: \"%s\"", String.join(" ", requestCmd)));
        // run command
        List<String> out = SendCommand(requestCmd);
        AssertValidCommand(out);
        Log.i("ATG", String.format("Output received from engine command: \"%s\"", String.join(" ", requestCmd)));
        return out;

    }

    public String Request(String[] requestCmd) throws InvalidCommandException {

        String outStr = String.join("\n", RequestAsList(requestCmd));
        return outStr;

    }

}

File: ATG/app/src/main/java/com/kuhakupixel/atg/backend/ACEBaseClient.kt
package com.kuhakupixel.atg.backend

import android.util.Log

abstract class ACEBaseClient {
    class InvalidCommandException(msg: String) : RuntimeException(msg)

    /**
     * specify how to send the command and get the output
     */
    abstract fun SendCommand(requestCmd: Array<String>): List<String>

    @Throws(InvalidCommandException::class)
    fun RequestAsList(requestCmd: Array<String>): List<String> {
        Log.i(
            "ATG",
            "Command to Engine: \"${requestCmd.joinToString(separator = " ")}\""
        )
        // run command
        val out = SendCommand(requestCmd)
        AssertValidCommand(out)
        Log.i(
            "ATG",
            "Output received from engine command: \"${requestCmd.joinToString(separator = " ")}\"",
        )
        return out
    }

    @Throws(InvalidCommandException::class)
    fun Request(requestCmd: Array<String>): String {
        return RequestAsList(requestCmd).joinToString(separator = "\n")
    }

    companion object {
        @Throws(InvalidCommandException::class)
        fun AssertValidCommand(out: List<String>) {
            if (out.size > 0) {
                if (out[0].equals("INVALID_COMMAND")) {
                    throw InvalidCommandException(
                        "Invalid Command: ${out.joinToString(separator = "\n")}"
                    )
                }
            }
        }
    }
}